<a href="https://colab.research.google.com/github/jasmine400/nlp-/blob/main/question_fa_embedding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
#access google drive
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [ ]:
#access google drive
%cd /content/gdrive/My\ Drive/Parsa
!ls

In [ ]:
# creating a data frame
df = pd.read_csv("export_dataframe6.csv")
print(df.head())

In [ ]:
print(len(df))

5065


In [ ]:
 question_ids= df._id.tolist()

In [ ]:
print(question_ids[1])

d9qvAn8B7wsuAN6fqXqe


In [ ]:
df = pd.DataFrame(question_ids, columns=['q_ar_id'])
print(df)

                   q_ar_id
0     N9quAn8B7wsuAN6fknlE
1     d9qvAn8B7wsuAN6fqXqe
2     utqwAn8B7wsuAN6fEXoH
3     vtqwAn8B7wsuAN6fFXo1
4     wtqwAn8B7wsuAN6fGXpH
...                    ...
5060  OtqiAn8B7wsuAN6fLmlQ
5061  3tqiAn8B7wsuAN6fd2kX
5062  _dqiAn8B7wsuAN6flmns
5063  UtqjAn8B7wsuAN6fnWsw
5064  gdqjAn8B7wsuAN6f0Wvy

[5065 rows x 1 columns]


In [ ]:
from google.colab import drive

drive.mount('/content/drive')


Mounted at /content/drive


In [ ]:
path = '/content/drive/My Drive/Parsa/q_ar_id.csv'

with open(path, 'w', encoding = 'utf-8-sig') as f:
  df.to_csv(f)


sava the id of the questions in each language

In [ ]:
#access google drive
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [ ]:
#access google drive
%cd /content/gdrive/My\ Drive/Parsa
!ls

/content/gdrive/My Drive/Parsa
creating_data_frame_log_0.ipynb  export_dataframe7.csv
creating_data_frame_v_0.ipynb	 export_dataframe8.csv
df_from_elk.ipynb		 export_dataframe9.csv
export_dataframe10.csv		 export_dataframefa.csv
export_dataframe11.csv		 find_optimal_num_clusters.ipynb
export_dataframe12.csv		 nba.csv
export_dataframe13.csv		 parsa_content.ipynb
export_dataframe14.csv		 prepared_df_for_embedding.csv
export_dataframe15.csv		 q_ar_id.csv
export_dataframe16.csv		 q_en_id.csv
export_dataframe17.csv		 q_fa_id.csv
export_dataframe18.csv		 q_tr_id.csv
export_dataframe2.csv		 question_fa_embedding.ipynb
export_dataframe3.csv		 reduction_q1.csv
export_dataframe4.csv		 reduction_q.csv
export_dataframe5.csv		 reduction_q.gsheet
export_dataframe6.csv		 topic_extraction.ipynb


In [ ]:
import pandas as pd

In [ ]:
#creat data frame
df = pd.read_csv('export_dataframefa.csv')

In [ ]:

df.head()

In [ ]:
df =df[['question','full_question','full_answer']]
print(len(df))

38857


In [ ]:
ids = pd.read_csv('q_fa_id.csv')

In [ ]:
df = df.join(ids)

In [ ]:
df.head()

In [ ]:
print(len(df))

38857


In [ ]:
#access count nan values in a specific column
df['full_answer'].isna().sum()

13504

we have 13504 nan values in the "full answer" column

In [ ]:
df['q&a'] = df['question'].map(str) + ',' + df['full_answer'].map(str)

In [ ]:
df.head()

In [ ]:
reduction=df[['index','q&a', 'q_fa_id','question']]

In [ ]:
df['index'] = range(0, len(df))

In [ ]:
df.head()

question  \
0             بی حس یا بی هوش کردن مجرم قبل از قصاص    
1                  عدم اطلاع شوهر از رجوع زن به بذل    
2             قصاص عضو ضارب در صورت بهبودی عضو مضروب   
3  رجوع زن به مهریه و عدم توانایی پرداخت مهریه تو...   
4          تملک ولی دم به اعضای بدن مجرم بعد از قصاص   

                                       full_question  \
0  کسى که قرار است کشته شود (تیرباران، حلق آویز و...   
1  این جانب با بذل یکصد عدد سکّه بهار آزادى از هم...   
2   اگر کسى انگشت یا دست کسى را عمداً قطع کند و م...   
3  اگر در ایّام طلاق خلع، زن به مهر خود رجوع کند ...   
4  هرگاه مجرم محکوم به قصاص اعدام شود، آیا ولى دم...   

                                         full_answer  Unnamed: 0  \
0  در اعدام یا همان قصاص نفس جایز نیست ولی در قطع...           0   
1  چنانچه در عدّه، رجوع به بذل خود کرده اید و به ...           1   
2  حقّ قصاص دارد ولى دوّمى نیز حق دارد انگشت خود ...           2   
3  اگر مقداری را بپردازد هر چند کم باشد، زوج حق ر...           3   
4                          ولىّ دم حقّى در آن ندارد.           4   

                q_fa_id                                                q&a  \
0  OdkpAn8B7wsuAN6febKj  بی حس یا بی هوش کردن مجرم قبل از قصاص ,در اعدا...   
1  OtkpAn8B7wsuAN6ferID   عدم اطلاع شوهر از رجوع زن به بذل ,چنانچه در ع...   
2  O9kpAn8B7wsuAN6ferIM  قصاص عضو ضارب در صورت بهبودی عضو مضروب,حقّ قصا...   
3  PNkpAn8B7wsuAN6ferJj  رجوع زن به مهریه و عدم توانایی پرداخت مهریه تو...   
4  PdkpAn8B7wsuAN6ferJr  تملک ولی دم به اعضای بدن مجرم بعد از قصاص,ولىّ...   

   index  
0      0  
1      1  
2      2  
3      3  
4      4

In [ ]:
from google.colab import drive

drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
path = '/content/drive/My Drive/parsa_content_based/reduction_q3.csv'

with open(path, 'w', encoding = 'utf-8-sig') as f:
  reduction.to_csv(f)

In [ ]:
path = '/content/drive/My Drive/Parsa/reduction_q1.csv'

with open(path, 'w', encoding = 'utf-8-sig') as f:
  reduction.to_csv(f)

In [ ]:
reduction.head()

index                                                q&a  \
0      0  بی حس یا بی هوش کردن مجرم قبل از قصاص ,در اعدا...   
1      1   عدم اطلاع شوهر از رجوع زن به بذل ,چنانچه در ع...   
2      2  قصاص عضو ضارب در صورت بهبودی عضو مضروب,حقّ قصا...   
3      3  رجوع زن به مهریه و عدم توانایی پرداخت مهریه تو...   
4      4  تملک ولی دم به اعضای بدن مجرم بعد از قصاص,ولىّ...   

                q_fa_id                                           question  
0  OdkpAn8B7wsuAN6febKj             بی حس یا بی هوش کردن مجرم قبل از قصاص   
1  OtkpAn8B7wsuAN6ferID                  عدم اطلاع شوهر از رجوع زن به بذل   
2  O9kpAn8B7wsuAN6ferIM             قصاص عضو ضارب در صورت بهبودی عضو مضروب  
3  PNkpAn8B7wsuAN6ferJj  رجوع زن به مهریه و عدم توانایی پرداخت مهریه تو...  
4  PdkpAn8B7wsuAN6ferJr          تملک ولی دم به اعضای بدن مجرم بعد از قصاص

In [ ]:
#access ondices of nan rows in a dataframe
#only "full answer column" have nan rows in dataframe
a=df[df['full_answer'].isnull()].index.tolist()
print(a)

[2074, 2075, 2076, 2077, 2078, 2079, 2080, 2081, 2082, 2083, 2084, 2085, 2086, 2087, 2088, 2089, 2090, 2091, 2092, 2093, 2094, 2095, 2096, 2097, 2102, 2106, 2109, 2110, 2111, 2112, 2113, 2114, 2115, 2118, 2119, 2121, 2122, 2123, 2124, 2125, 2126, 2127, 2132, 2133, 2134, 2135, 2136, 2139, 2140, 2141, 2142, 2143, 2144, 2145, 2146, 2148, 2149, 2150, 2151, 2152, 2153, 2154, 2155, 2156, 2157, 2159, 2167, 2170, 2171, 2174, 2175, 2176, 2177, 2178, 2179, 2180, 2181, 2182, 2183, 2184, 2185, 2186, 2190, 2191, 2192, 2193, 2194, 2195, 2201, 2202, 2203, 2204, 2205, 2206, 2207, 2208, 2209, 2210, 2211, 2212, 2213, 2215, 2216, 2217, 2218, 2219, 2220, 2221, 2222, 2224, 2225, 2227, 2228, 2229, 2230, 2231, 2233, 2235, 2236, 2237, 2238, 2239, 2240, 2242, 2244, 2246, 2247, 2248, 2249, 2253, 2254, 2255, 2256, 2257, 2258, 2260, 2262, 2263, 2264, 2265, 2266, 2267, 2270, 2271, 2272, 2273, 2274, 2275, 2276, 2278, 2279, 2280, 2281, 2282, 2283, 2284, 2285, 2288, 2290, 2292, 2293, 2294, 2296, 2298, 2299, 2300, 230

In [ ]:
print(len(df))

38857


In [ ]:
from google.colab import drive

drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
path = '/content/drive/My Drive/Parsa/prepared_df_for_embedding.csv'

with open(path, 'w', encoding = 'utf-8-sig') as f:
  df.to_csv(f)

create prper dataframe and concatenate "queation" & "full answer" columns in one column that named "q&a" column

In [ ]:
!nvidia-smi

Mon Feb 28 06:48:51 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla K80           Off  | 00000000:00:04.0 Off |                    0 |
| N/A   47C    P8    29W / 149W |      0MiB / 11441MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [ ]:
import torch
torch.cuda.is_available()
# Output would be True if Pytorch is using GPU otherwise it would be False.

True

In [ ]:
#check if gpu is on
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
print(f'device: {device}')

train_on_gpu = torch.cuda.is_available()

if not train_on_gpu:
    print('CUDA is not available.  Training on CPU ...')
else:
    print('CUDA is available!  Training on GPU ...')

device: cuda:0
CUDA is available!  Training on GPU ...


In [ ]:
#clear memory
import torch, gc

gc.collect()
torch.cuda.empty_cache()

!nvidia-smi

Mon Feb 28 06:49:16 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla K80           Off  | 00000000:00:04.0 Off |                    0 |
| N/A   46C    P8    29W / 149W |      3MiB / 11441MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [ ]:
!pip install -qU sentence-transformers
!pip install -qU wikipedia-api
!pip install -qU hazm
!pip install -qU clean-text[gpl]
!pip install -qU emoji

     |████████████████████████████████| 79 kB 3.6 MB/s 
     |████████████████████████████████| 3.5 MB 11.5 MB/s 
     |████████████████████████████████| 1.2 MB 45.2 MB/s 
     |████████████████████████████████| 67 kB 4.7 MB/s 
     |████████████████████████████████| 6.8 MB 35.1 MB/s 
     |████████████████████████████████| 895 kB 46.2 MB/s 
     |████████████████████████████████| 596 kB 40.9 MB/s 
     |████████████████████████████████| 316 kB 5.3 MB/s 
     |████████████████████████████████| 233 kB 39.4 MB/s 
     |████████████████████████████████| 1.4 MB 42.1 MB/s 
     |████████████████████████████████| 53 kB 1.4 MB/s 
     |████████████████████████████████| 174 kB 10.4 MB/s 
     |████████████████████████████████| 235 kB 41.1 MB/s 


In [ ]:
!mkdir resources
!wget -q "https://github.com/sobhe/hazm/releases/download/v0.5/resources-0.5.zip" -P resources
!unzip -qq resources/resources-0.5.zip -d resources

In [ ]:
!rm -rf /content/4ccae468eb73bf6c4f4de3075ddb5336
!rm -rf /content/preproc
!rm preprocessing.py utils.py
!mkdir -p /content/preproc
!git clone https://gist.github.com/4ccae468eb73bf6c4f4de3075ddb5336.git /content/preproc/
!mv /content/preproc/* /content/
!rm -rf /content/preproc

rm: cannot remove 'preprocessing.py': No such file or directory
rm: cannot remove 'utils.py': No such file or directory
Cloning into '/content/preproc'...
remote: Enumerating objects: 7, done.
remote: Total 7 (delta 0), reused 0 (delta 0), pack-reused 7
Unpacking objects: 100% (7/7), done.


In [ ]:
from transformers import BertConfig, BertTokenizer

from preprocessing import cleaning
from IPython import display

import numpy as np
import pandas as pd

import hazm
import requests
import time

import torch
from sentence_transformers import models, SentenceTransformer, util
from sklearn.cluster import KMeans

In [ ]:
def rtl_print(outputs, font_size="15px", n_to_br=False):
    outputs = outputs if isinstance(outputs, list) else [outputs]
    if n_to_br:
        outputs = [output.replace('\n', '<br/>') for output in outputs]

    outputs = [f'<p style="text-align: right; direction: rtl; margin-right: 10px; font-size: {font_size};">{output}</p>' for output in outputs]
    display.display(display.HTML(' '.join(outputs)))


def load_st_model(model_name_or_path):
    word_embedding_model = models.Transformer(model_name_or_path)
    pooling_model = models.Pooling(
        word_embedding_model.get_word_embedding_dimension(),
        pooling_mode_mean_tokens=True,
        pooling_mode_cls_token=False,
        pooling_mode_max_tokens=False)

    model = SentenceTransformer(modules=[word_embedding_model, pooling_model])
    return model

In [ ]:
# Load the Sentence-Transformer
embedder = load_st_model('m3hrdadfi/bert-fa-base-uncased-wikinli-mean-tokens')


Downloading:   0%|          | 0.00/519 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/621M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/404 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/1.14M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/112 [00:00<?, ?B/s]

In [ ]:
#confif bert model
config = BertConfig.from_pretrained(
    'm3hrdadfi/bert-fa-base-uncased-wikinli-mean-tokens', **{

    })
print(config.to_json_string())

{
  "_name_or_path": "HooshvareLab/bert-fa-base-uncased",
  "architectures": [
    "BertModel"
  ],
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "gradient_checkpointing": false,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "transformers_version": "4.16.2",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size": 100000
}



In [ ]:
#access google drive
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [ ]:
#access google drive
%cd /content/gdrive/My\ Drive/Parsa
!ls

/content/gdrive/My Drive/Parsa
creating_data_frame_log_0.ipynb  export_dataframe7.csv
creating_data_frame_v_0.ipynb	 export_dataframe8.csv
df_from_elk.ipynb		 export_dataframe9.csv
export_dataframe10.csv		 export_dataframefa.csv
export_dataframe11.csv		 find_optimal_num_clusters.ipynb
export_dataframe12.csv		 nba.csv
export_dataframe13.csv		 parsa_content.ipynb
export_dataframe14.csv		 prepared_df_for_embedding.csv
export_dataframe15.csv		 q_ar_id.csv
export_dataframe16.csv		 q_en_id.csv
export_dataframe17.csv		 q_fa_id.csv
export_dataframe18.csv		 q_tr_id.csv
export_dataframe2.csv		 question_fa_embedding.ipynb
export_dataframe3.csv		 reduction_q1.csv
export_dataframe4.csv		 reduction_q.csv
export_dataframe5.csv		 reduction_q.gsheet
export_dataframe6.csv		 topic_extraction.ipynb


In [ ]:
#creat data frame
df = pd.read_csv('reduction_q1.csv')

In [ ]:
df.head()

Unnamed: 0  ...               q_fa_id
0           0  ...  OdkpAn8B7wsuAN6febKj
1           1  ...  OtkpAn8B7wsuAN6ferID
2           2  ...  O9kpAn8B7wsuAN6ferIM
3           3  ...  PNkpAn8B7wsuAN6ferJj
4           4  ...  PdkpAn8B7wsuAN6ferJr

[5 rows x 3 columns]

In [ ]:
print(len(df))

38857


In [ ]:
corpus = list(df["q&a"])

In [ ]:
print(len(corpus))

38857


In [ ]:
# Load the Sentence-Transformer
embedder = load_st_model('m3hrdadfi/bert-fa-base-uncased-wikinli-mean-tokens')
corpus_embeddings = embedder.encode(corpus, show_progress_bar=True)

Batches:   0%|          | 0/1215 [00:00<?, ?it/s]

In [ ]:
from sklearn.cluster import KMeans
from sklearn import metrics
from scipy.spatial.distance import cdist
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
print(type(corpus_embeddings))
print(corpus_embeddings.shape)

<class 'numpy.ndarray'>
(38857, 768)


In [ ]:
distortions = []
inertias = []
mapping1 = {}
mapping2 = {}
K = range(1, 250)

for k in K:
	print(k)
	# Building and fitting the model
	kmeanModel = KMeans(n_clusters=k).fit(corpus_embeddings)
	kmeanModel.fit(corpus_embeddings)

	distortions.append(sum(np.min(cdist(corpus_embeddings, kmeanModel.cluster_centers_,
										'euclidean'), axis=1)) / corpus_embeddings.shape[0])
	inertias.append(kmeanModel.inertia_)

	mapping1[k] = sum(np.min(cdist(corpus_embeddings, kmeanModel.cluster_centers_,
								'euclidean'), axis=1)) / corpus_embeddings.shape[0]
	mapping2[k] = kmeanModel.inertia_

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40


In [ ]:
#create data frame
data =df[['q&a']]
data.head()

q&a
0  بی حس یا بی هوش کردن مجرم قبل از قصاص ,در اعدا...
1   عدم اطلاع شوهر از رجوع زن به بذل ,چنانچه در ع...
2  قصاص عضو ضارب در صورت بهبودی عضو مضروب,حقّ قصا...
3  رجوع زن به مهریه و عدم توانایی پرداخت مهریه تو...
4  تملک ولی دم به اعضای بدن مجرم بعد از قصاص,ولىّ...

In [ ]:
new_df=df.dropna()

In [ ]:
print(len(new_df))

38857


In [ ]:
df['question_id'] = mylist

question  ...                                        full_answer
0             بی حس یا بی هوش کردن مجرم قبل از قصاص   ...  در اعدام یا همان قصاص نفس جایز نیست ولی در قطع...
1                  عدم اطلاع شوهر از رجوع زن به بذل   ...  چنانچه در عدّه، رجوع به بذل خود کرده اید و به ...
2             قصاص عضو ضارب در صورت بهبودی عضو مضروب  ...  حقّ قصاص دارد ولى دوّمى نیز حق دارد انگشت خود ...
3  رجوع زن به مهریه و عدم توانایی پرداخت مهریه تو...  ...  اگر مقداری را بپردازد هر چند کم باشد، زوج حق ر...
4          تملک ولی دم به اعضای بدن مجرم بعد از قصاص  ...                          ولىّ دم حقّى در آن ندارد.

[5 rows x 4 columns]

In [ ]:
print(source[1])
print(type(source[1]))

{'language': 'fa', 'index': None, 'category': ['رجوع در طلاق خلع '], 'phrase': None, 'category_phrase': None, 'tags': ['شوهر', 'عدم اطلاع', 'بذل', 'رجوع زن'], 'question': ' عدم اطلاع شوهر از رجوع زن به بذل ', 'full_question': 'این جانب با بذل یکصد عدد سکّه بهار آزادى از همسرم مطلّقه خلعى شده ام، در زمان عدّه طىّ نامه کتبى به دفتردار محضر، مراتب رجوع به بذل را اعلام کردم، اکنون متوجّه شده ام که دفتردار رجوع قانونى را به علّت فراموشى در دفتر ثبت ننموده است، حکم رجوع چیست؟ آیا مستحقّ دریافت مهریّه هستم؟ ', 'answer': None, 'full_answer': 'چنانچه در عدّه، رجوع به بذل خود کرده اید و به شوهر اطّلاع داده اید، حق دارید بذل خود ر ابازپس گیرید و اگر خبر نداده اید عده گذشته است کافى نیست و هر گاه دفتردار متکفّل این کار شده و کوتاهى کرده است ضامن مى باشد. ', 'resource': None, 'translate': None, 'link': 'https://makarem.ir/main.aspx?typeinfo=21&lid=0&catid=45076&mid=256780', 'owner': 'makarem.ir'}
<class 'str'>


In [ ]:
import yaml
s = source[1]

ss=yaml.load(s)
print(type(ss))
print(ss)

<class 'dict'>
{'language': 'fa', 'index': 'None', 'category': ['رجوع در طلاق خلع '], 'phrase': 'None', 'category_phrase': 'None', 'tags': ['شوهر', 'عدم اطلاع', 'بذل', 'رجوع زن'], 'question': ' عدم اطلاع شوهر از رجوع زن به بذل ', 'full_question': 'این جانب با بذل یکصد عدد سکّه بهار آزادى از همسرم مطلّقه خلعى شده ام، در زمان عدّه طىّ نامه کتبى به دفتردار محضر، مراتب رجوع به بذل را اعلام کردم، اکنون متوجّه شده ام که دفتردار رجوع قانونى را به علّت فراموشى در دفتر ثبت ننموده است، حکم رجوع چیست؟ آیا مستحقّ دریافت مهریّه هستم؟ ', 'answer': 'None', 'full_answer': 'چنانچه در عدّه، رجوع به بذل خود کرده اید و به شوهر اطّلاع داده اید، حق دارید بذل خود ر ابازپس گیرید و اگر خبر نداده اید عده گذشته است کافى نیست و هر گاه دفتردار متکفّل این کار شده و کوتاهى کرده است ضامن مى باشد. ', 'resource': 'None', 'translate': 'None', 'link': 'https://makarem.ir/main.aspx?typeinfo=21&lid=0&catid=45076&mid=256780', 'owner': 'makarem.ir'}


In [ ]:
for i in range(len(source)):
  source[i]=yaml.load(source[i])